File: 3sum/3sum.go
// https://leetcode.com/problems/3sum/
// 33 ms 7.4 MB
package main

import "sort"

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	ans := [][]int{}
	for i := 0; i < len(nums)-2 && nums[i] <= 0; {
		for l, r := i+1, len(nums)-1; l < r; {
			cmp := nums[l] + nums[r] + nums[i]
			if cmp > 0 {
				// quick skip
				if nums[r]-nums[l+1] < cmp {
					break
				}
				r--
			} else if cmp < 0 {
				// quick skip
				if nums[l]-nums[r-1] > cmp {
					break
				}
				l++
			} else {
				ans = append(ans, []int{nums[i], nums[r], nums[l]})
				for l++; l <= r && nums[l] == nums[l-1]; l++ {
				}
				for r--; l <= r && nums[r] == nums[r+1]; r-- {
				}
			}
		}
		for i++; i < len(nums) && nums[i] == nums[i-1]; i++ {
		}
	}
	return ans
}

File: 3sum/3sum.js
// https://leetcode.com/problems/3sum/
// 174 ms 52.4 MB
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

var threeSum = function (nums) {
    nums.sort(function (a, b) {
        return a - b;
    });
    const res = [];
    const len = nums.length;
    let l, r;
    for (let i = 0; i < len - 2;) {
        l = i + 1;
        r = len - 1;
        while (l < r) {
            let sum = nums[i] + nums[l] + nums[r];
            if (sum > 0) {
                for (r--; r >= l && nums[r + 1] === nums[r]; r--);
            } else if (sum < 0) {
                for (l++; r >= l && nums[l - 1] === nums[l]; l++);
            } else {
                res.push([nums[i], nums[l], nums[r]])
                for (l++; r >= l && nums[l - 1] === nums[l]; l++);
                for (r--; r >= l && nums[r + 1] === nums[r]; r--);
            }
        }
        for (i++; i < len && nums[i] === nums[i - 1]; i++);
    }
    return res;
};